presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractRowBlock.java

```diff
@@ -131,11 +134,12 @@ public Block copyRegion(int position, int length)
             newBlocks[i] = getFieldBlocks()[i].copyRegion(startFieldBlockOffset, fieldBlockLength);
         }
 
-        int[] newOffsets = new int[length + 1];
-        for (int i = 1; i < newOffsets.length; i++) {
-            newOffsets[i] = getFieldBlockOffset(position + i) - startFieldBlockOffset;
+        int[] newOffsets = sliceOffsets(getFieldBlockOffsets(), position + getOffsetBase(), length);
+        boolean[] newRowIsNull = compactBooleanArray(getRowIsNull(), position + getOffsetBase(), length);
+
+        if (Arrays.equals(newBlocks, getFieldBlocks()) && newOffsets == getFieldBlockOffsets() && newRowIsNull == getRowIsNull()) {
```

Comment:
I'm not comfortable with depending on "Block.equals(Object)" being not implemented. Please add a `arraySame` function.